tests: Update DevSim Tests

jeremyk-lunarg opened the pull request:
Updating devsim tests with a new Python test script, new tests for
portability, and new comparison files.

lunarpapillo commented on tests/devsim_layer_tests.py:
```diff
@@ -0,0 +1,181 @@
+    
+    vulkaninfo_path = args.vulkan_tools_install_dir
+    
+    if sys.platform.startswith("win32"):
+        vulkaninfo_path = os.path.join(vulkaninfo_path, "bin", "vulkaninfo.exe")
```
I'm not a fan of OS-dependent code; but I don't have an easy way to make this sequence OS-independent, so I'm only grumbling about it.

lunarpapillo commented on tests/devsim_layer_tests.py:
```diff
@@ -0,0 +1,181 @@
+        vulkaninfo_path = os.path.join(vulkaninfo_path, "bin", "vulkaninfo.exe")
+    elif sys.platform.startswith("linux"):
+        vulkaninfo_path = os.path.join(vulkaninfo_path, "bin", "vulkaninfo")
+    elif sys.platform.startswith("darwin"):
+        vulkaninfo_path = os.path.join(vulkaninfo_path, "vulkaninfo", "vulkaninfo")
```
Is this correct on Mac?

lunarpapillo commented on tests/devsim_layer_tests.py:
```diff
@@ -0,0 +1,181 @@
+    
+    test_in_files = ["devsim_test2_in1.json", "devsim_test2_in2.json", "devsim_test2_in3.json", "devsim_test2_in4.json", "devsim_test2_in5.json"]
+    
+    test_dict["BasicTest"] = TestFramework(test_in_files, BasicTest(TestEnvironmentSettings(), "test1_vulkaninfo_out.json", "devsim_test2_gold.json"), vulkaninfo_path)
+    test_dict["PortabilityExtensionPresentEmulationOffTest"] = TestFramework(["devsim_dummy_in.json"], PortabilityExtensionPresentTest(TestEnvironmentSettings(), "port_extent_test_vulkaninfo_out.json", None), vulkaninfo_path)
+    test_dict["PortabilityExtensionPresentEmulationOnTest"] = TestFramework(["devsim_dummy_in.json"], PortabilityExtensionPresentTest(TestEnvironmentSettings(portability=1), "port_extent_test_vulkaninfo_out.json", None), vulkaninfo_path)
+    test_dict["PortabilityNonEmulatedTest"] = TestFramework(["portability_test.json"], PortabilityTest(TestEnvironmentSettings(), "port_test_vulkaninfo_out.json", "portability_test_gold.json"), vulkaninfo_path)
+    test_dict["PortabilityEmulatedTest"] = TestFramework(["portability_test.json"], PortabilityTest(TestEnvironmentSettings(portability=1), "port_test_vulkaninfo_out.json", "portability_test_gold.json"), vulkaninfo_path)
```
As above, it's easier and clearer to:
```
test_dict = {
    "BasicTest": TestFramework(...),
    "PortabilityExtensionPresentEmulationOffTest": TestFramework(...),
    ...
}
```

lunarpapillo commented on tests/devsim_layer_tests.py:
```diff
@@ -0,0 +1,181 @@
+    for k in test_dict:
+        if k in test_blacklist:
+            continue
+        print("RUNNING TEST")
+        print(k)
```
Suggest: `print("RUNNING TEST", k)`

lunarpapillo commented on tests/devsim_layer_tests.py:
```diff
@@ -0,0 +1,181 @@
+            continue
+        print("RUNNING TEST")
+        print(k)
+        print("{}".format("PASS" if test_dict[k].runAndCheck() else "FAIL"))
```
Don't need the `"{}".format()` construction.  Suggest: 
```
print("PASS" if text_dict[k].runAndCheck() else "FAIL")
```

lunarpapillo commented on tests/devsim_test2_gold.json:
```diff
@@ -58,61 +32,61 @@
-    "variableMultisampleRate": 154,
-    "vertexPipelineStoresAndAtomics": 126,
-    "wideLines": 116
+    "alphaToOne": 1,
```
Were the old values IDs of some kind?  I don't understand why these were changed.

jeremyk-lunarg replied:
These values are `VkBool32` which can technically be any 32-bit integer.

The original test made these values weird, ridiculous numbers to show that devsim could simulate a driver with values that didn't make any sense.

Due to updates in `vulkaninfo`, these values are forced to `0` or `1` in output, so in order for the tests to pass now, the gold files had to be updated to the new `vulkaninfo` output.

lunarpapillo commented on tests/portability_test.json:
```diff
@@ -0,0 +1,17 @@
+    "VkPhysicalDevicePortabilitySubsetPropertiesKHR": {
+        "minVertexInputBindingStrideAlignment": 8
+    }
+}
```
Seems to be missing a terminal newline, where almost all the other files have one.

lunarpapillo commented on tests/portability_test_gold.json:
```diff
@@ -0,0 +1,22 @@
+    "VkPhysicalDevicePortabilitySubsetPropertiesKHR": {
+        "minVertexInputBindingStrideAlignment": 8
+    }
+}
```
Seems to be missing a termianl newline, where almost all of the other files have one.

lunarpapillo reviewed (changes requested):
Lots of good work here!  So I have lots of suggestions. :-)

It's not clear what version of Python is being targeted here.  Whatever version is being used, the corresponding version of `pylint` should also be used to validate the Python.  (`pylint` will catch many of the things I caught, and more.)

It's not clear to me where these tests are going to be run.  Because they are in the **VulkanTools** repository, I presume these are developer tests and/or Travis/AppVeyor tests?  If they're intended to run in internal LunarG CI, there's a lot more that needs to be done, to match the internal test harness.

The commit message is insufficient to justify all the changes here (in particular, the changes to values in the `.json` files) - some description of why that was done should appear.

jeremyk-lunarg commented:
@lunarpapillo, I am intending for this to be used by internal CI.

What additional changes do I need to make for it to be compatible with the internal test harness?

lunarpapillo commented:
> @lunarpapillo, I am intending for this to be used by internal CI.
> 
> What additional changes do I need to make for it to be compatible with the internal test harness?

Internal tests all live in **LunarG/VulkanTests**, and use the utilities and test classes there, which provide a lot of features, including:
-  full logging, to help determine what happened when a test failed
- capture of all stdout and stderr for all external operations
- saved test artifacts/results for later review
- OS-independent operation, requiring all OS-dependent operations to be abstracted out and placed in OS-dependent classes, to allow a test to run on all supported OSes without having to change if a new OS is added
- etc...

The options here are:
- move test scripts entirely into **VulkanTests**
- create a new **VulkanTests** script that invokes the project scripts as a module
- create a new **VulkanTests** script that invokes the project scripts as an external executable

The choice depends on whether the test scripts are intended to be available to developers, Travis, or AppVeyor; or whether they're just intended for LunarG internal CI.

For more on the test harness: https://github.com/LunarG/VulkanTests/wiki/Test-Suite-Architecture

lunarpapillo commented:
Woot!  Looks good, but I didn't review the changes in `devsim_layer_tests.py` closely, as they look similar/identical to what they were during our live review.

Did you have to change this file much?  If you could direct me to the locations that changed, I'll review those more thoroughly.

Also, it looks like your Ubuntu builds are failing, likely because a file is missing from this commit...

jeremyk-lunarg commented:
@lunarpapillo It's passing now!

The changes I made to `devsim_layer_tests.py` are in commit 18bd6e34b15347ed27342bd308e2b1284910c145.

In summary:
- I had to add another flag that was added to Devsim recently.
- I fixed a test that had improper settings.
- I removed memory flag modification from one of the tests.  This will be tested in a future test.

I also made some changes in commit ab3abca5191b9d6583615afd3f699427b602e11d.

In summary:
- The tests need to be run in the same directory as the test files.  I added an argument to allow a user to set the run directory from the terminal.

It was a quick change to make the GitHub Actions work and could be made cleaner if this is going to be used as a module, but I'll let you be the judge of that.

lunarpapillo commented on tests/devsim_layer_tests.py:
```diff
@@ -363,10 +363,22 @@ def main():
     args = parser.parse_args()
 
     vulkaninfo_path = args.vulkan_tools_install_dir
+    working_directory = args.directory
+
+    starting_working_directory = os.getcwd()
+
+    if working_directory is not None:
+        working_directory = os.path.abspath(working_directory)
```
It's easier to provide the value as a default argument:
```
parser.add_argument(
    "--dir",
    help="Specify the directory to run the tests in.  Defaults to current working directory.",
    dest="directory",
    default=os.getcwd()
)
```
Then after `parse_args()`:
```
working_directory = os.path.abspath(args.directory)
```

jeremyk-lunarg replied:
Good suggestion!
Made the change and pushed it up.  Will merge after cloud CI passes.

lunarpapillo reviewed (approved):
Looks great!  I suggest one change, but the code is clear and understandable as is, so if you don't want to make the suggested change I still approve. :+1:
